src/dtx/dtx_rpc.c

```diff
@@ -606,164 +572,143 @@ dtx_classify_one(struct ds_pool *pool, daos_handle_t tree, d_list_t *head, int *
 }
 
 static int
-dtx_rpc_prep(struct ds_cont_child *cont, d_list_t *head, struct btr_root *tree_root,
-	     daos_handle_t *tree_hdl, struct dtx_req_args *dra, ABT_thread *helper,
-	     struct dtx_id dtis[], struct dtx_entry **dtes, daos_epoch_t epoch,
-	     uint32_t count, int opc, int *committed)
+dtx_rpc_prep(struct ds_cont_child *cont,d_list_t *dti_list,  struct dtx_entry **dtes,
+	     uint32_t count, int opc, daos_epoch_t epoch, d_list_t *cmt_list,
+	     d_list_t *abt_list, d_list_t *act_list, struct dtx_common_args *dca)
 {
-	d_rank_t	my_rank;
-	uint32_t	my_tgtid;
-	int		rc;
+	struct dtx_req_args	*dra;
+	int			 rc = 0;
 
-	D_INIT_LIST_HEAD(head);
-	dra->dra_future = ABT_FUTURE_NULL;
-	crt_group_rank(NULL, &my_rank);
-	my_tgtid = dss_get_module_info()->dmi_tgt_id;
+	memset(dca, 0, sizeof(*dca));
 
-	/* Use helper ULT to handle DTX RPC if there are enough helper XS. */
-	if (dss_has_enough_helper() &&
-	    (dtes[0]->dte_mbs->dm_tgt_cnt - 1) * count >= dtx_rpc_helper_thd) {
-		struct dtx_helper_args	*dha = NULL;
-
-		D_ALLOC_PTR(dha);
-		if (dha == NULL)
-			return -DER_NOMEM;
-
-		dha->dha_cont = cont;
-		dha->dha_head = head;
-		dha->dha_tree_root = tree_root;
-		dha->dha_tree_hdl = tree_hdl;
-		dha->dha_dra = dra;
-		dha->dha_dtes = dtes;
-		dha->dha_epoch = epoch;
-		dha->dha_count = count;
-		dha->dha_opc = opc;
-		dha->dha_committed = committed;
-		dha->dha_rank = my_rank;
-		dha->dha_tgtid = my_tgtid;
-
-		rc = dss_ult_create(dtx_rpc_helper, dha, DSS_XS_IOFW,
-				    my_tgtid, DSS_DEEP_STACK_SZ, helper);
-		if (rc != 0) {
-			D_FREE(dha);
-		} else if (dtis != NULL) {
-			int	i;
+	D_INIT_LIST_HEAD(&dca->dca_head);
+	dca->dca_tree_hdl = DAOS_HDL_INVAL;
+	dca->dca_epoch = epoch;
+	dca->dca_count = count;
+	crt_group_rank(NULL, &dca->dca_rank);
+	dca->dca_tgtid = dss_get_module_info()->dmi_tgt_id;
+	dca->dca_cont = cont;
+	dca->dca_helper = ABT_THREAD_NULL;
+	dca->dca_dtes = dtes;
 
-			for (i = 0; i < count; i++)
-				dtis[i] = dtes[i]->dte_xid;
-		}
+	dra = &dca->dca_dra;
+	dra->dra_future = ABT_FUTURE_NULL;
+	dra->dra_cmt_list = cmt_list;
+	dra->dra_abt_list = abt_list;
+	dra->dra_act_list = act_list;
+	dra->dra_opc = opc;
+	uuid_copy(dra->dra_po_uuid, cont->sc_pool->spc_pool->sp_uuid);
+	uuid_copy(dra->dra_co_uuid, cont->sc_uuid);
+
+	if (dti_list != NULL) {
+		d_list_splice(dti_list, &dca->dca_head);
+		D_INIT_LIST_HEAD(dti_list);
 	} else {
-		rc = dtx_rpc_internal(cont, head, tree_root, tree_hdl, dra, dtis, dtes, epoch,
-				      count, opc, committed, my_rank, my_tgtid);
+		if (count > 1) {
+			D_ALLOC_ARRAY(dca->dca_dtis, count);
+			if (dca->dca_dtis == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+		} else {
+			dca->dca_dtis = &dca->dca_dti_inline;
+		}
 	}
 
+	/* Use helper ULT to handle DTX RPC if there are enough helper XS. */
+	if (dss_has_enough_helper())
+		rc = dss_ult_create(dtx_rpc_helper, dca, DSS_XS_IOFW, dca->dca_tgtid,
+				    DSS_DEEP_STACK_SZ, &dca->dca_helper);
+	else
+		rc = dtx_rpc_internal(dca);
```

Comment:
just a note that even when there is no helper XS, DSS_XS_IOFW will be executed on neighbor's vos main XS. (not saying that is better but just FYI).

Reply:
Yes, the main purpose of using helper XS for DTX RPC is mainly to avoid blocking under heavy system load. If no helper XS configured, then using neighbor main XS may be also blocked.